### ISSUE TITLE
cmd/compile: wrong evaluation order for map assignment [Go1.8]

### ISSUE THREAD

**Issue by randall77** [golang/go#22881]
```
package main

func main() {
	m := map[int]int{}
	defer func() {
		recover()
		println(len(m))
	}()
	m[5] = m[5] + *p
}

var p *int
```
This should print 0, not 1.  The nil ptr panic should happen before the insert into m.
It prints 0 up to go 1.7, but starting at go 1.8 it prints 1.
Same thing happens if I use `s[0]` on a nil slice instead of `*p` on a nil pointer.

The bug goes away if I do `tmp := m[5] + *p; m[5] = tmp`.

I'm going to label this as a backportable 1.9.3 issue so we can discuss (and maybe also 1.8.6, but github won't let me multi-milestone an issue), but I kind of lean toward just fixing this in 1.10.  It's pretty esoteric, survived more than a release cycle without a bug report, and it is easy to work around.

It's also possible I'm not reading the spec correctly, and this is allowed behavior. @griesemer 

Doesn't seem to be SSA related - the AST already has the ops out of order on input to SSA.

**Comment by mvdan** [comment]
Slightly simpler version, as a playground link: https://play.golang.org/p/5hasX7yOX1

**Comment by griesemer** [comment]
Looks like a bug to me. The spec says that the right-hand side is first fully evaluated before any assignments take place. A nil-pointer deref causes a runtime panic similar to calling panic() which (immediately) terminates the surrounding function.

**Comment by mdempsky** [comment]
Yeah, this is probably an order.go issue. We can't call mapassign until we know the RHS will evaluate without panicking.

At least OIND, OINDEX, ODIV, and OMOD can panic, but are currently allowed after the mapassign call.

**Comment by randall77** [comment]
Reopening for backport consideration (both 1.9.3 and 1.8.6).

**Comment by go101** [comment]
This is not only for map
```
package main

var x int
func f2() *int {
  println("bye") // this line shouldn't be output.
  return &x
}

func main() {
        var p *int
        *f2() = *p
}
```

**Comment by go101** [comment]
however, if we view left `m[key]`as `*GetMapValueAddr(key)`, then the example is the first comment may be also viewed as not a bug.

**Comment by go101** [comment]
Similar to this problem: https://groups.google.com/forum/#!topic/golang-nuts/J3kXXZivlHA

**Comment by go101** [comment]
Then this
```
package main

import "fmt"

func main() {
  var m = map[int]int{}
  
  defer func() {
    fmt.Println(recover())
    fmt.Println(len(m)) // 0
  }()
  var p *int
  m[2], *p = 1, 2
}
```

The spec says `m[2]=1` happens before `*p=2`, so the print result should be 1 instead.

**Comment by ianlancetaylor** [comment]
Let's take detailed discussion of order of evaluation away from this issue, which is about a specific bug.  Thanks.

**Comment by ianlancetaylor** [comment]
Closing proposed backport per discussion in #23005.  (This remains fixed in the upcoming 1.10 release.)
